### PR TITLE
[Concurrency] Implement parsing and semantic analysis of await operator

### PR DESCRIPTION
**Pull request by DougGregor**
Similar to `try`, await expressions have no specific semantics of their
own except to indicate that the subexpression contains calls to `async`
functions, which are suspension points. In this design, there can be
multiple such calls within the subexpression of a given `await`.

Note that we currently use the keyword `__await` because `await` in
this position introduces grammatical ambiguities. We'll wait until
later to sort out the specific grammar we want and evaluate
source-compatibility tradeoffs. It's possible that this kind of prefix
operator isn't what we want anyway.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
Scavenged from https://github.com/apple/swift/pull/11501, which was good code that just needed a little love after sitting idle for 3 years.

File: lib/Sema/TypeCheckError.cpp [hunk 1]
```diff
@@ -1505,25 +1552,45 @@ class CheckErrorCoverage : public ErrorHandlingWalker<CheckErrorCoverage> {
...
+    if (classification.isAsync()) {
+      // Remember that we've seen an async call.
+      Flags.set(ContextFlags::HasAnyAsyncSite);
+      
+      // Diagnose async calls that are outside of an await context.
+      if (!Flags.has(ContextFlags::IsAsyncCovered)) {
+        SourceRange highlight;
+        
+        // Generate more specific messages in some cases.
+        if (auto e = dyn_cast_or_null<ApplyExpr>(E.dyn_cast<Expr*>()))
```
Comment (LucianoPAlmeida) [inline lib/Sema/TypeCheckError.cpp:1582]: Nit: Not important, but question is if could this be a place to use `getAsExpr<ApplyExpr>(E)`?

Comment (DougGregor) [reply lib/Sema/TypeCheckError.cpp:1582]: Probably. I don't think that function existed when this was originally written ;)

Comment (theblixguy) [reply lib/Sema/TypeCheckError.cpp:1582]: @LucianoPAlmeida Unfortunately it cannot be used here as it's defined in ConstraintSystem.

Comment (LucianoPAlmeida) [reply lib/Sema/TypeCheckError.cpp:1582]: Ah true @theblixguy, so maybe move to a more general header file (as you suggested) would be nice to allow use in other places :)

**Pull request merged**